### DISCUSSION TITLE
timeout error 400.

### DISCUSSION THREAD

**Discussion by kabapy** [pocketbase/pocketbase#4612] [Q&A] [answered]
Repeating part of the filter query (unintentionally) may cause pocketbase to time out giving error 400 after long wait.

I experienced this issue with a large db 500k+. It has simple posts, tags collections

Example
```
tags~"oxs3b259cm0a7xu" || tags~"oxs3b259cm0a7xu"
```
Tags is a relations field many-to-many
Notice it is the same tag

A larger query filter may caused pocketbase to crash on my system multiple times (not sure it was caused by this)

My implementation is on Nodejs

**Comment by ganigeorgiev** [comment]
The query timeout is expected if it runs more than 30s (you can change the default with the `--queryTimeout` flag).

If you need help, please provide details about your collections structure (you can export it from the _Admin UI > Settings > Export collections_) and what API rules or other filters you have.

Additionally, if `tags` in the above is a multiple `relation` field a more efficient query could be to use `tags.id ?= "yourid"`.

**Reply by ganigeorgiev** [reply]
To help with the debugging you can also start the executable with `--dev` and paste the generated query.

**Reply by kabapy** [reply]
I did try increasing timeout to 60. Not working.
I will try the dev flag

**Reply by ganigeorgiev** [reply]
@kabapy I'm getting tired repeating this, but please export your collections configuration as json (_Admin UI > Settings > Export collections_) because otherwise there is no way for me to reproduce your issue or help you.

Coming from your recent feature request in https://github.com/pocketbase/pocketbase/issues/4609, I suspect that this could be caused by an index that trips the query planner to take a more unefficient path similar to https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045.

**Reply by kabapy** [reply]
I apologize for my shot answer. I was out of the office

Here is my schema.

[pb_schema.json](https://github.com/pocketbase/pocketbase/files/14721944/pb_schema.json)


The posts table has 500k rows. I usually get 40 seconds for simple filter like
`tags.id?="oxs3b259cm0a7xu"`

Thank you for your patience

**Reply by ganigeorgiev** [reply]
Could you start the executable with `--dev` and post the query that is printed in the console?

Also why both `tags` and `posts` has relations with each other? Are both relations filled in your data set?

Do you also need the combined `niche, used_in_projects` index? As noted in https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045 this may not always be a good idea as it could trip the query planner to take a more inefficient path, so try to remove it if you don't need the index.

**Reply by kabapy** [reply]
I will do run that as soon a sget back to the office.
The schema is little messy. The app is in active development.

Yes, both relations fields are currently populated, we may drop the tags table side.

Yes, that is the troublesome query, I need to filter by 3 filters niche and used_in_projects and tags

I created this index after having slow performance and timeout on my table.

**Comment by ganigeorgiev** [comment, marked as answer]
I've spent some time trying to create a test environment and debug the issue and I suspect that the problematic query in your case is the `COUNT` which unfortunately doesn't allow much room for improvements even if you play around with the indexes.

### Test env setup

First here is the seeder dummy script I've used to populate my test database based on your provided collections shema:

<details>
<summary><b>Seeder script</b></summary>

```js
import PocketBase from "pocketbase";

const pb = new PocketBase("http://127.0.0.1:8090");

pb.autoCancellation(false);

const now = (new Date()).toISOString().replace("T", " ")

// 500k tags
const tags = [];
for (let i = 0; i < 500; i++) {
    const promises = [];
    for (let j = 0; j < 1000; j++) {
        const id = (i + "_" + j).padStart(15, '0');
        promises.push(pb.collection("tags").create({
            "id":  id,
            "tag": "test_" + id,
        }));
        tags.push(id);
    }
    await Promise.all(promises);
}

// 500k media
const media = [];
for (let i = 0; i < 500; i++) {
    const promises = [];
    for (let j = 0; j < 1000; j++) {
        const id = (i + "_" + j).padStart(15, '0');
        promises.push(pb.collection("media").create({
            "id":       id,
            "url":      "https://google.com",
            "type":     j % 2 == 0 ? "VIDEO" : "IMAGE",
            "approved": j % 2 == 0,
            "reviewed": j % 2 == 0,
        }));
        media.push(id);
    }
    await Promise.all(promises);
}

// 500k projects
const projects = [];
for (let i = 0; i < 500; i++) {
    let promises = [];
    for (let j = 0; j < 1000; j++) {
        const id = (i + "_" + j).padStart(15, '0');
        promises.push(pb.collection("projects").create({
            "id":                   id,
            "name":                 "name_" + id,
            "social_network":       j % 2 ? "Facebook" : "Telegram",
            "niche":                "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec mi lectus, pharetra molestie finibus sed, iaculis ac sem. Donec bibendum, arcu at ultricies pulvinar, elit magna sollicitudin diam, eget hendrerit ligula nunc non nulla. Duis faucibus orci ut venenatis vestibulum. Morbi ultrices nec est vitae viverra. Vestibulum aliquam ligula sed neque molestie pretium. Aliquam erat volutpat. Vestibulum at laoreet elit. Integer tellus justo, euismod vitae lacinia non, viverra eget ipsum. Praesent gravida massa id mi fringilla elementum. Suspendisse eu interdum lacus. Donec pharetra, elit id cursus iaculis, odio felis fermentum libero, in pretium orci metus ut lacus.",
            "credentials":          {"a": "abc", "b": [1,2,3], "c": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."},
            "frequency_in_minutes": 10,
            "last_publish":         now,
            "active":               j % 2 == 0,
            "marketing_messages":   {"a": "abc", "b": [1,2,3], "c": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."},
            "media_type":           j % 2 ? "IMAGE" : "VIDEO",
            "config":               {"a": "abc", "b": [1,2,3], "c": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."},
            "preferred_tags":       {"a": "abc", "b": [1,2,3], "c": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."},
            "social_template1":     "social_template1_" + id,
            "social_template2":     "social_template2_" + id,
            "evergreen":            j % 2 == 0,
            "type":                 "Message",
        }));
        projects.push(id);
    }
    await Promise.all(promises);
}

// 500k posts
for (let i = 0; i < 500; i++) {
    let promises = [];
    for (let j = 0; j < 1000; j++) {
        const id = (i + "_" + j).padStart(15, '0');
        promises.push(pb.collection("posts").create({
            "id":               id,
            "title":            "title_" + id,
            "body":             "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec mi lectus, pharetra molestie finibus sed, iaculis ac sem. Donec bibendum, arcu at ultricies pulvinar, elit magna sollicitudin diam, eget hendrerit ligula nunc non nulla. Duis faucibus orci ut venenatis vestibulum. Morbi ultrices nec est vitae viverra. Vestibulum aliquam ligula sed neque molestie pretium. Aliquam erat volutpat. Vestibulum at laoreet elit. Integer tellus justo, euismod vitae lacinia non, viverra eget ipsum. Praesent gravida massa id mi fringilla elementum. Suspendisse eu interdum lacus. Donec pharetra, elit id cursus iaculis, odio felis fermentum libero, in pretium orci metus ut lacus.",
            "source_url":       "https://google.com",
            "publish_date":     now,
            "scraper_name":     "test_" + id,
            "media_type":       j % 2 == 0 ? "IMAGE" : "VIDEO",
            "niche":            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec mi lectus, pharetra molestie finibus sed, iaculis ac sem. Donec bibendum, arcu at ultricies pulvinar, elit magna sollicitudin diam, eget hendrerit ligula nunc non nulla. Duis faucibus orci ut venenatis vestibulum. Morbi ultrices nec est vitae viverra. Vestibulum aliquam ligula sed neque molestie pretium. Aliquam erat volutpat. Vestibulum at laoreet elit. Integer tellus justo, euismod vitae lacinia non, viverra eget ipsum. Praesent gravida massa id mi fringilla elementum. Suspendisse eu interdum lacus. Donec pharetra, elit id cursus iaculis, odio felis fermentum libero, in pretium orci metus ut lacus.",
            "metdata":          {"a": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec mi lectus, pharetra molestie finibus sed, iaculis ac sem. Donec bibendum, arcu at ultricies pulvinar, elit magna sollicitudin diam, eget hendrerit ligula nunc non nulla. Duis faucibus orci ut venenatis vestibulum. Morbi ultrices nec est vitae viverra. Vestibulum aliquam ligula sed neque molestie pretium. Aliquam erat volutpat. Vestibulum at laoreet elit. Integer tellus justo, euismod vitae lacinia non, viverra eget ipsum. Praesent gravida massa id mi fringilla elementum. Suspendisse eu interdum lacus. Donec pharetra, elit id cursus iaculis, odio felis fermentum libero, in pretium orci metus ut lacus. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.", "b": [1,2,3], "c": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."},
            "tags":             randomItems(tags),
            "media":            randomItems(media),
            "used_in_projects": randomItems(projecs),
        }));
    }
    await Promise.all(promises);
}

function randomItems(arr, max = 100) {
    const result = [];

    max = arr.length > max ? max : arr.length;

    for (let i = 0; i < max; i++) {
        result.push(arr[Math.floor(Math.random() * arr.length)]);
    }

    return result;
}
```
</details>

The dummy script creates 500K records for each `tags`, `media`, `project` and `posts` collection.
Each `posts` record has 100 random relation references respectively for each relation field (`tags`, `media`, `used_in_projects`).

My test request that triggers the timeout is the following one:

```js
await pb.collection("posts").getList(1, 10, {
    filter: "tags.id ?= 'YOUR_TAG_ID'"
})
```

The reason for this as mentioned in the beginning is the `COUNT` query that we perform by default in order to populate the `totalItems` and `totalPages` response fields.

`COUNT` queries (and aggregations in general) are slow when performed over large datasets with rows with large text/json data because such queries often require a full table scan and at some point or another it has to read/copy the rows data.

The issue is not PocketBase, nor SQLite only specific and you'll probably experience the same with other more traditional DB .

The easiest way to rule out PocketBase issue is to execute the minimum working query directly against the SQLite file (for example with the `sqlite3` cli tool as we cannot make it faster than this):

```sql
-- sqlite3 pb_data/data.db

SELECT COUNT(DISTINCT posts.id)
FROM posts
LEFT JOIN json_each(posts.tags) je
WHERE je.value = "YOUR_TAG_ID";
```
The above query takes locally ~35-40sec. The exact same query executed with PocketBase using the CGO driver is identical.


### Possible workarounds

Unfortunately I'm not sure how to optimize the above query other than avoid executing it entirely but there are couple workarounds that you can try:

1. Use `skipTotal=1` and `tags:each`

    By specifying the `skipTotal=1` query parameter it will instruct the action to skip the count query entirely.

    By using `tags:each` instead of `tags.id` in the filter we can avoid the extra relation join and directly search within the json array value (_the performance gain is small but it helps a little; in future versions we'll do this transparently behind the scenes, aka. `tags.id` and `tags:each` will generate the same query if there is no other `tags.*` field constraint_)

    ```js
    // skipTotal:1 is automatically set for getFullList and getFirstListItem requests
    await pb.collection("posts").getList(1, 10, {
        filter: "tags:each ?= 'YOUR_TAG_ID'",
        skipTotal: 1,
    })
    ```
    The above generates the following raw query that executes for ~1s:
    ```sql
    [1082.00ms] SELECT DISTINCT `posts`.* FROM `posts` LEFT JOIN json_each(CASE WHEN json_valid([[posts.tags]]) THEN [[posts.tags]] ELSE json_array([[posts.tags]]) END) `posts_tags_je` WHERE [[posts_tags_je.value]] = '00000000499_822' LIMIT 10
    ```
    The response takes additional ~5s which is mostly because of the large data and the not very efficient json serialization and allocations around the public exportable fields. I've already optimized the last part while working on the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355). I'll try this weekend to see if I can apply some of the optimizations in v0.22 as the refactoring could take a while.

2. Specify the `fields` parameter and/or create a [View collection](https://pocketbase.io/docs/collections/#view-collection) with only the fields you need. For example requests against the following View collection query return almost immediately:
   ```sql
   SELECT id, title, body FROM posts
   ```

3. Move the relation fields in a junction collection(s), for example `postTag` which could contain only 2 relation fields to the the `posts` and `tags` collections. **This requires major changes to your data write and read flow so evaluate carefully.**

**Reply by kabapy** [reply]
thank you for you effort an time and your detailed explanation. i did run the query with --dev here is the results

the console logs
[report.txt](https://github.com/pocketbase/pocketbase/files/14727928/report.txt)

the query link is
http://127.0.0.1:8094/api/collections/posts/records?page=1&perPage=1&filter=niche%20%3D%20%22HEALTH%22%20%26%26%20used_in_projects%20!~%20%22h0wzo507hsya7zp%22%20%20%26%26%20(tags.id%3F%3D%22dtnpfqu3jvaqebq%22%20%7C%7C%20tags.id%3F%3D%22wnwhwjdhs9pcvro%22%20%7C%7C%20tags.id%3F%3D%22o0rqzgxp18ds70d%22%20%7C%7C%20tags.id%3F%3D%22bi0w4di9qb48mdg%22%20%7C%7C%20tags.id%3F%3D%22dm3ajwbjjlomkm6%22%20%7C%7C%20tags.id%3F%3D%22qikw3msb9r0n29n%22%20%7C%7C%20tags.id%3F%3D%22dyq1zln98qbfnd8%22%20%7C%7C%20tags.id%3F%3D%220mztebivjv9a0j0%22%20%7C%7C%20tags.id%3F%3D%22n2tlm0l5qxh3zuj%22%20%7C%7C%20tags.id%3F%3D%229m8gr75kgk8ccwz%22%20%7C%7C%20tags.id%3F%3D%228y9q1sgeuxv59jv%22%20%7C%7C%20tags.id%3F%3D%221i9zhxug4aporji%22%20%7C%7C%20tags.id%3F%3D%229tt9t3njhjnly8x%22%20%7C%7C%20tags.id%3F%3D%22pkck5x9037fiu7a%22%20%7C%7C%20tags.id%3F%3D%22mpvbxfs1ks0j2iq%22%20%7C%7C%20tags.id%3F%3D%22njkisdednjjexas%22%20%7C%7C%20tags.id%3F%3D%228680siz63y33h34%22%20%7C%7C%20tags.id%3F%3D%22nb097p0jmnllzdc%22%20%7C%7C%20tags.id%3F%3D%22tw64dt5a7xfi0ft%22%20%7C%7C%20tags.id%3F%3D%22g776u78mld2fvzk%22%20%7C%7C%20tags.id%3F%3D%22rdbdjyrfnsir5bq%22%20%7C%7C%20tags.id%3F%3D%2250b4lt3aade58s5%22%20%7C%7C%20tags.id%3F%3D%22auyiy433v4nd84z%22%20%7C%7C%20tags.id%3F%3D%22oy4pnzncb053sbo%22%20%7C%7C%20tags.id%3F%3D%22pyc9xtu1ta7wxm1%22%20%7C%7C%20tags.id%3F%3D%22c2u7d7k6jzlqyxx%22%20%7C%7C%20tags.id%3F%3D%22c3gobkhppohajvx%22%20%7C%7C%20tags.id%3F%3D%22crmyp32bxo28ps2%22%20%7C%7C%20tags.id%3F%3D%22dtc4edb4rsi5xox%22%20%7C%7C%20tags.id%3F%3D%224gefguocww6yra7%22%20%7C%7C%20tags.id%3F%3D%22si6kwvh79x6vfk8%22%20%7C%7C%20tags.id%3F%3D%225y3xlnhtyvqwrjs%22%20%7C%7C%20tags.id%3F%3D%22hz6ci31rvmjauoj%22%20%7C%7C%20tags.id%3F%3D%22tvsxk64tspsmpuu%22%20%7C%7C%20tags.id%3F%3D%22kzdkrmklt0u5me9%22%20%7C%7C%20tags.id%3F%3D%222tvej9rtio40zzy%22%20%7C%7C%20tags.id%3F%3D%2271nnzg8no6h2508%22%20%7C%7C%20tags.id%3F%3D%22zloamkq9nusbsy0%22%20%7C%7C%20tags.id%3F%3D%22du854lm5k8sxmt6%22%20%7C%7C%20tags.id%3F%3D%22d65usa6m9x2a3pl%22%20%7C%7C%20tags.id%3F%3D%22dwy8kd51uih6mt0%22%20%7C%7C%20tags.id%3F%3D%22ybtauaww159reye%22%20%7C%7C%20tags.id%3F%3D%22yna068umhxcnxge%22%20%7C%7C%20tags.id%3F%3D%22z3fwdd9mdwg5laj%22%20%7C%7C%20tags.id%3F%3D%22fknbg0242j9nhoh%22%20%7C%7C%20tags.id%3F%3D%22b5mj9vq9z8x0964%22%20%7C%7C%20tags.id%3F%3D%22zm2z0gifa9vvtoq%22%20%7C%7C%20tags.id%3F%3D%22pbo767zkiqm6a70%22%20%7C%7C%20tags.id%3F%3D%22pijvtm1c951y1s1%22%20%7C%7C%20tags.id%3F%3D%22ffquu010p89hq2p%22%20%7C%7C%20tags.id%3F%3D%22ijeabuelzwr3eug%22%20%7C%7C%20tags.id%3F%3D%228cm1s6sv5gm2er1%22%20%7C%7C%20tags.id%3F%3D%22rr6efpda5jfi7o7%22%20%7C%7C%20tags.id%3F%3D%22li0mm2fxmsx4lnm%22%20%7C%7C%20tags.id%3F%3D%22j5vzkpd3w8e0qnk%22%20%7C%7C%20tags.id%3F%3D%22gamz12u9sw6oqwq%22%20%7C%7C%20tags.id%3F%3D%22uzklu2wp7e8x90x%22%20%7C%7C%20tags.id%3F%3D%228ui77tpl0eosnxz%22%20%7C%7C%20tags.id%3F%3D%22iljocu08gomjp0e%22%20%7C%7C%20tags.id%3F%3D%22tf3520t6hy1s3iy%22%20%7C%7C%20tags.id%3F%3D%22w7fuoy3twltlw66%22%20%7C%7C%20tags.id%3F%3D%22mjbx4oal85j0rqw%22%20%7C%7C%20tags.id%3F%3D%225tv5pxpp9kj19wg%22%20%7C%7C%20tags.id%3F%3D%22g347e8jv846orar%22%20%7C%7C%20tags.id%3F%3D%22svnjd0pcke59w3r%22%20%7C%7C%20tags.id%3F%3D%22tle17uko4pa74hy%22%20%7C%7C%20tags.id%3F%3D%223u8o0umyf5h8mvn%22%20%7C%7C%20tags.id%3F%3D%22ao93ujrjazlsytq%22%20%7C%7C%20tags.id%3F%3D%22z42wdialnsp7e8z%22%20%7C%7C%20tags.id%3F%3D%22lpnl3nf12vsnw76%22%20%7C%7C%20tags.id%3F%3D%224b112sp7r4drdqr%22%20%7C%7C%20tags.id%3F%3D%22mmgb1ickqxy8fxv%22%20%7C%7C%20tags.id%3F%3D%225pi1yik9li070oy%22%20%7C%7C%20tags.id%3F%3D%22ybzuikkzws95xcg%22%20%7C%7C%20tags.id%3F%3D%22jce9nzh1jt58bqx%22%20%7C%7C%20tags.id%3F%3D%22tshftaanp55zc9x%22%20%7C%7C%20tags.id%3F%3D%220fdqqlqhm3c1250%22%20%7C%7C%20tags.id%3F%3D%22iitizzemix33qaw%22%20%7C%7C%20tags.id%3F%3D%22igc4ohv4rjgn0j7%22%20%7C%7C%20tags.id%3F%3D%22maonxg3sq9tifom%22%20%7C%7C%20tags.id%3F%3D%225x329u6m1um5mqg%22%20%7C%7C%20tags.id%3F%3D%22aqyj17idc2gwkch%22%20%7C%7C%20tags.id%3F%3D%22qf8j5zexhcwxu24%22%20%7C%7C%20tags.id%3F%3D%2286uvk7ydspji6y8%22%20%7C%7C%20tags.id%3F%3D%22eqdxolmsorpam10%22%20%7C%7C%20tags.id%3F%3D%22e1e4v3kqiqbqazg%22%20%7C%7C%20tags.id%3F%3D%22rlea3b8hqb3z8rc%22%20%7C%7C%20tags.id%3F%3D%22nhc55jp4a0win28%22%20%7C%7C%20tags.id%3F%3D%22luf8ekdalpie25v%22%20%7C%7C%20tags.id%3F%3D%22fej41ae97w43sj8%22%20%7C%7C%20tags.id%3F%3D%22xgpur1nswtacmpi%22%20%7C%7C%20tags.id%3F%3D%22bbxmd9epnwcse45%22%20%7C%7C%20tags.id%3F%3D%22oudsztsp1a9d3pl%22%20%7C%7C%20tags.id%3F%3D%22x30xt1g9eusunev%22)&sort=%40random

**Reply by ganigeorgiev** [reply]
Yes, the logs also confirms that it is the COUNT query. Adding `&skipTotal=1` to your url would definitely help speed up the execution. If the performance after doing this is not satisfiable, the only solution I can think of is to restructure your collection columns and keep the large json/text ones separately.

I also forgot to mention that running `sqlite3 pb_data/data.db 'PRAGMA optimize'` could also help in your case. In future PocketBase versions we'll run it periodically as this is also what [SQLite recommends](https://www.sqlite.org/pragma.html#pragma_optimize) for long running processes.
If you want to do that on your own, below is one possible example with the JS `pb_hooks`:

```js
// run after app start
onAfterBootstrap(() => {
    $app.dao().db().newQuery("PRAGMA optimize").execute()
})

// run every hour
cronAdd("pragma_optimize", "0 * * * *", () => {
    $app.dao().db().newQuery("PRAGMA optimize").execute()
})
```

**Comment by kabapy** [comment]
Thank you for your advice. I tried several of your suggestions.

It finally works . the query time get down to few seconds. 
I restructured my code to create one big query that search both posts and tags
using `:each` helped but the decisive factor was indeed ` skipTotal: 1`.

Also, i discovered that using a sort option `sort: "@random"` made things worse. i used the getList to get items then randomized the list in node.js

Quick questions:
1. why did you implemented relation in JSON fields. isn't using a many to many intermediate table a better more compliant way to manage relations?
2. are you using internal caching mechanisms, if so, any info about the limits of the cache

thank you very much for your help. I wish you the best

**Reply by ganigeorgiev** [reply]
1. We provide both options and you can manage the relations in separate table if you want to do it the same way as in a more traditional db setup. The "multiple" `relation` field is used primarily for simplicity to avoid the need of creating and managing a junction table. Note that the `JSON` column itself (technically it is a regular `TEXT` field) is not the reason why the query is slow. It is the large amount of data in each `posts` row that slows it down. SQLite recently [introduced `JSONB`](https://sqlite.org/draft/json1.html#jsonbx) that could help eventually with the size and operations over large json fields but we haven't migrated to it yet.

2. We don't do any explicit caching out of the box at the moment (_SQLite could have some cache for the parsing of frequently executed SQL statements but it is done transparently for us_).

**Reply by kabapy** [reply]
i see, any permanent suggestions for the 
> the large amount of data in each posts row that slows it down

that i can implement. i don't want to face a slowdown with larger data volumes

**Reply by ganigeorgiev** [reply]
Unfortunately there is no "one-size-fit-all" type of solution and it will depend on your application data and desired data access patterns.

The general recommendations I guess could be:
- find the troublesome query and see if an index could help but avoid specifying combined non-unique indexes if you are not sure that you need them as it can sometimes cause more harm than good (_see https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045_); the easiest way to debug the query is using the `sqlite3` cli tool and running `EXPLAIN QUERY PLAN yourQuery` (in future PocketBase version we'll try to add indexes suggestions directly from the Admin UI).
- avoid having collections with large number of fields
- if possible store large text/json values that doesn't need to be queried/joined frequently in a separate collection
- occasionally run `sqlite3 pb_data/data.db 'PRAGMA optimize'` and/or `sqlite3 pb_data/data.db 'PRAGMA analyze'` (the latter is slow)
